src/util/LanguageUtil.ts

```diff
@@ -0,0 +1,125 @@
+import Bundles from 'util/Bundles';
+import config from 'config/config.json';
+import BrowserFeatures from './BrowserFeatures';
+
+import { LanguageCallback, LanguageCode } from 'types';
+import Util from './Util';
+
+/**
+ * Utility that gets a list of language tag preferences based on the widget properties
+ * and browser settings.
+ * 
+ * NOTE: The language tag and language are related but not necessarilty the same.
+ * The language tag specifies the language and region, and may be used for formatting dates
+ * and currency, while the language is just the language code used for translations.
+ * 
+ * @see https://en.wikipedia.org/wiki/IETF_language_tag
+ * 
+ * @param {string | LanguageCallback | undefined} language 
+ * @param {string[]} supportedLanguages
+ * @returns {string[]}
+ */
+export const getLanguageTags = (language: string | LanguageCallback | undefined, supportedLanguages: string[]): string[] => {
+  // Get the user's preferred languages from the browser API
+  const userLanguages = BrowserFeatures.getUserLanguages().map((userLanguage: string) => {
+    // Map "simple" language codes to their full locale equivalents to match the
+    // expected format for these specific languages.
+    if (userLanguage === 'nl') {
+      return 'nl-nl';
+    }
+    if (userLanguage === 'pt') {
+      return 'pt-br';
+    }
+    return userLanguage.toLowerCase();
+  });
+
+  const languagesMap = new Map<string, Set<string>>();
+
+  const addLanguageTag = (languageTag?: string): void => {
+    if (!languageTag) {
+      return;
+    }
+    // only get the simple language code, i.e. 'en', 'ja', 'zh'
+    const [lang] = languageTag.split('-');
+
+    if (languagesMap.has(lang)) {
+      const values = languagesMap.get(lang) ?? new Set<string>();
+      values.add(languageTag)
+      languagesMap.set(lang, values);
+    } else {
+      languagesMap.set(lang, new Set<string>([languageTag]));
+    }
+  };
+
+  const supportedUserLanguages = userLanguages.filter((userLanguage: string) => {
+    // remove any user languages tags that do not match a supported language
+    return supportedLanguages.some((supportedLang: string) => userLanguage.startsWith(supportedLang.toLowerCase()));
```

Comment:
does `userLanguage` also need to be lower cased before comparing?

Reply:
`userLanguages` is always lowercase -- we do `.toLowerCase()` in the code above from L24-L33

Reply:
ah got it thank you!